nipype/interfaces/base/support.py

```diff
@@ -9,185 +9,166 @@
 """
 from __future__ import (print_function, division, unicode_literals,
                         absolute_import)
-from builtins import object, str
+from builtins import object
 
-import os
-from copy import deepcopy
+from ... import __version__
 
-from ... import logging
-from ...utils.misc import is_container
-from ...utils.filemanip import md5, to_str, hash_infile
-iflogger = logging.getLogger('nipype.interface')
 
-
-class NipypeInterfaceError(Exception):
+class NipypeInterfaceError(RuntimeError):
     """Custom error for interfaces"""
+    pass
 
-    def __init__(self, value):
-        self.value = value
 
-    def __str__(self):
-        return '{}'.format(self.value)
+class InterfaceRuntime(object):
+    """A class to store runtime details of interfaces
 
+    Elements can be set at initialization:
 
-class Bunch(object):
-    """Dictionary-like class that provides attribute-style access to it's items.
+    >>> rt = InterfaceRuntime(cmdline='/bin/echo', returncode=0)
+    >>> rt.returncode == 0
+    True
 
-    A `Bunch` is a simple container that stores it's items as class
-    attributes.  Internally all items are stored in a dictionary and
-    the class exposes several of the dictionary methods.
+    Unset elements are initialized to ``None``:
 
-    Examples
-    --------
-    >>> from nipype.interfaces.base import Bunch
-    >>> inputs = Bunch(infile='subj.nii', fwhm=6.0, register_to_mean=True)
-    >>> inputs
-    Bunch(fwhm=6.0, infile='subj.nii', register_to_mean=True)
-    >>> inputs.register_to_mean = False
-    >>> inputs
-    Bunch(fwhm=6.0, infile='subj.nii', register_to_mean=False)
+    >>> rt.cwd is None
+    True
 
-    Notes
-    -----
-    The Bunch pattern came from the Python Cookbook:
+    And any element can be set by attribute:
 
-    .. [1] A. Martelli, D. Hudgeon, "Collecting a Bunch of Named
-           Items", Python Cookbook, 2nd Ed, Chapter 4.18, 2005.
+    >>> rt.cwd = '/scratch/workflow'
+    >>> rt.cwd
+    '/scratch/workflow'
 
-    """
+    The structure raises ``AttributeError`` if trying to accessing
+    an invalid attribute:
 
-    def __init__(self, *args, **kwargs):
-        self.__dict__.update(*args, **kwargs)
+    >>> rt.nonruntime
+    Traceback (most recent call last):
+        ...
+    AttributeError: 'InterfaceRuntime' object has no attribute 'nonruntime'
 
-    def update(self, *args, **kwargs):
-        """update existing attribute, or create new attribute
 
-        Note: update is very much like HasTraits.set"""
-        self.__dict__.update(*args, **kwargs)
+    For Nipype to work, it is fundamental that runtime objects
+    can be pickled/unpickled:
 
-    def items(self):
-        """iterates over bunch attributes as key, value pairs"""
-        return list(self.__dict__.items())
+    >>> import pickle
+    >>> pickleds = pickle.dumps(rt)
+    >>> newrt = pickle.loads(pickleds)
+    >>> newrt
+    Bunch(cmdline='/bin/echo', cwd='/scratch/workflow', returncode=0)
```

Comment:
Why `Bunch` and not `InterfaceRuntime`?

Reply:
That is for backwards compatibility. If I change Bunch I'm afraid I'll need to update all the tests. But I can make that extra mile.